src/rebar_prv_upgrade.erl

```diff
@@ -111,12 +111,32 @@ do_(State) ->
                       [element(1,Lock) || Lock <- Locks],
                       [rebar_app_info:name(App) || App <- rebar_state:lock(State5)]
                      ),
-                    rebar_prv_lock:do(State5);
+                    State6 = maybe_ignore_checkouts(State5, Checkouts),
+                    rebar_prv_lock:do(State6);
                 _ ->
                     Res
             end
     end.
 
+-spec maybe_ignore_checkouts(rebar_state:t(), [binary()]) -> rebar_state:t().
+maybe_ignore_checkouts(State, Checkouts) ->
+    OldLocks = rebar_state:get(State, {locks, default}, []),
+    OldLockNames = [element(1,L) || L <- OldLocks],
+    case OldLockNames -- Checkouts of
+        %% If there aren't any checkout dependencies in the old lock info,
+        %% we can safely "ignore" them to avoid a duplicated message from
+        %% `rebar_prv_lock:info_checkout_deps/1`.
+        OldLockNames ->
+            DepsIgnoringCheckout =
+                [case rebar_app_info:is_checkout(Dep) of
+                        true -> rebar_app_info:is_checkout(Dep, false);
```

Comment:
I believe this will now hide status of whether a dep is a checkout dep to any provider that relies on locking (i.e. compiling).  This may break a lot of stuff.

Reply:
Yeah, I thought so too. If the state that `lock` uses has no checkout apps, it will definitely be passed onto other providers that `lock` calls/relies on.

Thanks for bringing it up, because for a second I forgot that although the code executed by `lock` will not have any issues, the provider returns its own state for others to use.

Would it make sense then to, in line 115, match on `{ok, State7} = rebar_prv_lock:do(State6)` and patch again the state by marking again as checkouts accordingly the dependencies?

Reply:
That might work. It could possibly be simpler to just keep a quick cache entry about it. The rebar agent does it internally by carrying explicit state:
https://github.com/erlang/rebar3/blob/c102379385013896711bba3969f280f851c67cc7/src/rebar_agent.erl#L166-L172
The config handler uses the app env values to do it:
https://github.com/erlang/rebar3/blob/ba50a0828eba37d386c5f82cf1814d4519d66f64/src/rebar_config.erl#L97-L112

The pdict could work in more restrained cases.

Reply:
Yeah using app environment is a nice thing that rebar3 uses in some features. I do not think, however, it would translate well into removing a duplicated message :/

As for the cache, the idea is good, but given that the `lock` provider does not modify the list of dependencies, the simplest approach IMO is to keep said list in a variable and after calling that provider from `upgrade`, setting the list of dependencies of the new state to that. No need to introduce a record/cache to keep track of the list when the use case is this minimal, although I will keep it in mind for other providers :)